### PR TITLE
 exp run: track files for skipped/run-cached stages

### PR DESCRIPTION
**Pull request by pmrowla**
* [x] ❗ I have followed the [Contributing to DVC](https://dvc.org/doc/user-guide/contributing/core) checklist.

* [x] 📖 If this PR requires [documentation](https://dvc.org/doc) updates, I have created a separate PR (or issue, at least) in [dvc.org](https://github.com/iterative/dvc.org) and linked it here.

Thank you for the contribution - we'll try to review it as soon as possible. 🙏

Fixes https://github.com/iterative/dvc/issues/9781

### CODE REVIEW

File: dvc/repo/experiments/executor/base.py [hunk 1]
```diff
@@ -546,6 +535,17 @@ def reproduce(
...
+    @staticmethod
+    def _repro_and_track(stage: "Stage", **kwargs) -> Optional["Stage"]:
+        from dvc.repo.reproduce import _reproduce_stage
+        from dvc.stage.utils import _get_stage_files
+
+        ret = _reproduce_stage(stage, **kwargs)
+        if not kwargs.get("dry") and (paths := _get_stage_files(stage)):
+            logger.debug("Staging stage-related files: %s", paths)
+            stage.repo.scm_context.add(paths)
+        return ret
```
Comment (pmrowla) [inline dvc/repo/experiments/executor/base.py:547]: `stage.reproduce()` returns None for skipped (unchanged run-cached) stages, and when the entire pipeline is skipped `dvc.reproduce()` ends up returning an empty list. In `exp run` we still need to force git-tracking the `dvc.yaml` and `dvc.lock` for those skipped stages.

**Comment by skshetry** [comment]
@pmrowla, what do you think of changing the return type of `dvc.reproduce()`?

**Comment by skshetry** [comment]
> > @pmrowla, what do you think of changing the return type of `dvc.reproduce()`?
> 
> I thought about having it return the tuple of `([reproduced], [skipped])` stages but wasn't sure if that would be useful anywhere else outside of this use case

I think it should be something like following now, after `--keep-going` and `--ignore-errors` implementation.
```python
@dataclass
class Result:
    reproduced: List["Stage"] = field(default_factory=list)
    unchanged: List["Stage"] = field(default_factory=list)
    failed: List["Stage"] = field(default_factory=list)
    skipped: List["Stage"] = field(default_factory=list)
```

I think we are long due for changing return type of `repro`. We have been hacking just to not change the return type.

But I guess adjusting tests will be painful, so I'm happy with this too for now. But we'll have to change the return type at some point.

**Comment by skshetry** [comment]
I am also okay with introducing a `post_run_hook` as that's what this seems to be about, and will avoid wrapping `_reproduce_stage`.

**Comment by skshetry** [comment]
Btw, what should be the behaviour here if `dvc exp run` fails? Should it still stage files from stages that were run? Previous implementation did not do that, this one does.

**Comment by daavoo** [comment]
I agree changing the return type would be cleaner but no strong opinion or need to block the P.R.

**Comment by pmrowla** [comment]
> Btw, what should be the behaviour here if `dvc exp run` fails? Should it still stage files from stages that were run? Previous implementation did not do that, this one does.

If it fails, we don't make a commit and everything gets `git reset` to the state prior to the experiment anyways, so it doesn't make a difference whether or not we stage anything right now

**Comment by pmrowla** [comment]
I think we're all in agreement that the return type for `reproduce()` will need to be reworked going forward, but I think that's beyond the scope of this bug fix, especially since it will require updating a lot of tests

Going to go ahead and merging this as-is for now

**Pull request merged**